### DISCUSSION TITLE
A modular network stack

### DISCUSSION THREAD

**Discussion by TomzBench** [libp2p/rust-libp2p#2490] [General]
Hello,

I cam across here via: https://libp2p.io/ while looking for a cross platform device discovery solution (mdns/bonjour/zeroconf).

The libp2p website kind of implies I could pull in a mdns lib, but in the example looks like I need the whole libp2p stack. Digging further, the mdns implementation is non standard, and there is some discussion to address that https://github.com/libp2p/rust-libp2p/issues/2394 and https://github.com/libp2p/rust-libp2p/issues/1815.

But even the current implementation seems to hard code the host name, meaning my discovery solution could bump into libp2p nodes - and the api doesn't allow me to distinguish myself from other libp2p nodes.

So I am curious what is the goal for libp2p.

Is there going to be a refactor where a user could just pull in only mdns if that's all that they need?
Is this library intended for general networking applications or is this specific for a single purpose related to web3?

**Comment by mxinden** [comment]
Hi @TomzBench,

>  and the api doesn't allow me to distinguish myself from other libp2p nodes.

You will discover other libp2p nodes, though you can decide who to connect to.

> Is there going to be a refactor where a user could just pull in only mdns if that's all that they need?

You can use the mdns protocol only. In other words you don't have to use kademlia, gossipsub, ping, ...

>  Is this library intended for general networking applications or is this specific for a single purpose related to web3?

Very much general networking application. Usecases range from file-sharing, over blockchains to industry applications.

**Reply by TomzBench** [reply]
hello @mxinden 

> You can use the mdns protocol only. In other words you don't have to use kademlia, gossipsub, ping, ...

But in order to use mdns protocol only, I need to initialize a swarm, which needs a peer id which needs crypto. (and therefor pulls in a lot of unnecessary code, and so is not very modular.)

> You will discover other libp2p nodes, though you can decide who to connect to.

But how would I decide? The mdns implementation only exposes the known addresses of remote peers, with no means to discovery what kind of services are hosted by the remote peer. It seems this implementation is assuming it is designed for one purpose and not very applicable for general use cases.

**Reply by mxinden** [reply]
Your problem statements above are correct.

I guess our definition of "general" differs here. I can understand that libp2p is not general enough for some use-cases.

Unfortunately, I don't think it makes sense for libp2p to become more general in the direction you are looking for. E.g. I do think libp2p's authentication mechanism should be a required core building block that every other libp2p protocol can assume to be present and thus depend on.

**Reply by TomzBench** [reply]
I understand, thanks for clarity on libp2p's scope
